feat: add dunder method for math operations on Axes Size divider

paquiteau opened the pull request:
<!--
Thank you so much for your PR!  To help us review your contribution, please check
out the development guide https://matplotlib.org/devdocs/devel/index.html
-->

## PR summary

Closes #27930. This adds __dunder__ methods (mul, and divide notably) to make interacting with `axes_grid1.Size` object better. 


<!-- Please provide at least 1-2 sentences describing the pull request in detail
(Why is this change required?  What problem does it solve?) and link to relevant
issues and PRs.

Also please summarize the changes in the title, for example "Raise ValueError on
non-numeric input to set_xlim" and avoid non-descriptive titles such as "Addresses
issue #8576".
-->


The original code for Size object interaction is not tested, so I don't know how to proceed to add some test for these changes. Overall, the direct usage of `Size` objects is quite niche, so I seek guidance as well for the documentation (maybe a new example ?) 


## PR checklist
<!-- Please mark any checkboxes that do not apply to this PR as [N/A].-->

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [ ] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [ ] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [ ] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [N/A ] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

Thanks in advance!

<!--We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.-->

timhoffm commented on lib/mpl_toolkits/axes_grid1/axes_size.py:
```diff
@@ -19,12 +19,27 @@ class _Base:
     def __rmul__(self, other):
         return Fraction(other, self)
 
+    def __mul__(self, other):
+        return Fraction(self, other)
```
Looking at the implementation of `Fraction` this does not work. I'm sorry, this is not well documented, but the Fraction signature is `def __init__(self, fraction : numbers.Real, ref_size : _Base)`, meaning a fraction of the given size. So `0.3 * mysize` is 30% of `mysize`. If anything, you'd have to return `Fraction(other, self)` as well. But since this is basically a scaling factor, it may be better to only accept the factor on the left.

timhoffm commented on lib/mpl_toolkits/axes_grid1/axes_size.py:
```diff
@@ -19,12 +19,27 @@ class _Base:
+    def __mul__(self, other):
+        return Fraction(self, other)
+
+    def __div__(self, other):
+        return Fraction(self, 1/other)
```
Similar to above `mysize / 2` is half of `mysize`. Here, it clearly does not make sense to divide something by a size. So this should not be added.

timhoffm commented on lib/mpl_toolkits/axes_grid1/axes_size.py:
```diff
@@ -19,12 +19,27 @@ class _Base:
             return Add(self, Fixed(other))
 
+    def __radd__(self, other):
+        return Add(self, Fixed(other))
```
I would add a comment, explaining the asymmetry compared to `__add__`
```suggestion
        # other cannot be a _Base instance, because A + B would trigger
        # A.__add__(B) already
        return Add(self, Fixed(other))
```

timhoffm commented on lib/mpl_toolkits/axes_grid1/axes_size.py:
```diff
@@ -19,12 +19,27 @@ class _Base:
+
+    def __sub__(self, other):
+        if isinstance(other, _Base):
+            return Add(self, -other)
```
This will not work as `_Base` does not define `__neg__`.

paquiteau commented:
Thanks for the early review!

 Following your comments, I added some type checking for scaling operations. 
A clear definition (and documentation) of the behavior could be specified. From what I understood from the code, `Size` can be scaled (e.g. multiply by a `Real` number) and added or substracted together (basically all operations defined for a linear vector space).

I do think that allowing multiplication on the right (or div) is nice to have (especially division): Typically to split an axis equally in n part: 

```python
divider.set_horizontal([Size.AxesX(ax) / n for _ in range(n)]
```

tacaswell commented:
Rather than `TypeError` we should be returning https://docs.python.org/3/library/constants.html#NotImplemented to play-nice with the  fallback rules and letting the interpreter untimely raise the error.

timhoffm reviewed (commented):
Would be nice to have better documentation and tests on `Size`, but given that we don't have that in the current code, I'm not requiring it for this PR.

anntzer commented on lib/mpl_toolkits/axes_grid1/axes_size.py:
```diff
@@ -17,14 +21,41 @@ class (or others) to determine the size of each Axes. The unit
+        return Add(self, Fixed(other))
+
+    def __sub__(self, other):
+        if isinstance(other, _Base):
```
I think `__sub__` can just be defined (in all cases, including the exception-raising one) as `self + (-other)`?

anntzer reviewed (approved):
Just a minor nit.

anntzer commented on lib/mpl_toolkits/axes_grid1/axes_size.py:
```diff
@@ -17,14 +21,39 @@ class (or others) to determine the size of each Axes. The unit
+    def __sub__(self, other):
+        if isinstance(other, (_Base, Real)):
+            return self + (-other)
+        return NotImplemented
```
The point is that you don't even need the isinstance check: it will be done by `__add__`.  Literally `def __sub__(self, other): return self + (-other)` should be enough.

anntzer commented on lib/mpl_toolkits/axes_grid1/axes_size.py:
```diff
@@ -17,14 +21,39 @@ class (or others) to determine the size of each Axes. The unit
+            return NotImplemented
         return Fraction(other, self)
 
+    def __div__(self, other):
```
Likewith this is `return (1/other) * self` with no other check.

anntzer commented on lib/mpl_toolkits/axes_grid1/axes_size.py:
```diff
@@ -17,14 +21,39 @@ class (or others) to determine the size of each Axes. The unit
 
 class _Base:
     def __rmul__(self, other):
+        if not isinstance(other, Real):
```
Likewise this is `return self * other` with no other check.

anntzer commented:
The pre-commit failure is real (typo; "subtraction", not "substraction"); also, can you squash the commits?

anntzer commented:
Thanks for the PR!
